### PR TITLE
Support different topologies

### PR DESCRIPTION
**Pull request by shai-doron**
issue #1

### CODE REVIEW

File: network-config-analyzer/NetworkConfigQuery.py [hunk 1]
```diff
@@ -510,6 +511,107 @@ def exec(self):
...
+                        protocol_name = ConnectionSet.protocol_number_to_name(protocol)
+                        port_range = str(entry.removed.allowed_protocols[protocol])
+                        if port_range not in removed[protocol_name]:
+                            removed[protocol_name][port_range] = []
+                        removed[protocol_name][port_range].append((entry.from_ep, entry.to_ep))
+            explanation = f'Added connections:\n{self.pretty_print_diff(added)}\n'
+            explanation += f'Removed connections:\n{self.pretty_print_diff(removed)}\n'
+
+            return QueryAnswer(False, 'Base branch and analyzed commit are not semantically equivalent.', explanation)
+        return QueryAnswer(True, 'Base branch and analyzed commit are semantically equivalent.')
```
Comment (zivn) [inline network-config-analyzer/NetworkConfigQuery.py:614]: These are GitSecure-specific messages. Should be replaced with something more generic

File: tests/example_policies/testSemanticDiff/with-different-topologies-scheme.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+# testing with networkPolicyLists (for backward compatibility)
+# can use only global namespaceList and podList resources
+networkPolicyLists:
+  - name: no_policy
+    networkPolicyList: []
+    expected_warnings: 0
+
+  - name: policy
+    namespaceList: ../tests/example_podlist/ns_list.json
```
Comment (zivn) [inline tests/example_policies/testSemanticDiff/with-different-topologies-scheme.yaml:12]: This is currently not allowed. Does that mean that `networkPolicyLists` and `networkConfigList` are synonyms?

File: tests/example_podlist/bank_pod_list.json [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+              "running": {
+                "startedAt": "2018-03-25T11:24:34Z"
+              }
+            },
+            "lastState": {
+
+            },
+            "ready": true,
+            "restartCount": 0,
+            "image": "registry.ng.net/demo_k8s/cognetive:latest",
```
Comment (zivn) [inline tests/example_podlist/bank_pod_list.json:611]: I think `cognetive` is kind of IBM-internal. Maybe rename.

File: network-config-analyzer/NetworkConfigQuery.py (hunk 1 above)
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:614]: done

File: tests/example_policies/testSemanticDiff/with-different-topologies-scheme.yaml (hunk 2 above)
Comment (shai-doron) [reply tests/example_policies/testSemanticDiff/with-different-topologies-scheme.yaml:12]: Yes, good point. Therefore I suggest to clean up and remove deprecated `networkPolicyLists`

File: tests/example_podlist/bank_pod_list.json (hunk 3 above)
Comment (shai-doron) [reply tests/example_podlist/bank_pod_list.json:611]: done

File: tests/example_policies/testSemanticDiff/with-different-topologies-scheme.yaml (hunk 2 above)
Comment (shai-doron) [reply tests/example_policies/testSemanticDiff/with-different-topologies-scheme.yaml:12]: Deprecated networkPolicyLists is removed

**Comment by shai-doron** [comment]
@zivn - deprecated networkPolicyLists has been removed with all unit tests pass successfully. Pls review changes and approve for merge.

File: network-config-analyzer/SchemeRunner.py [hunk 4]
```diff
@@ -47,29 +47,41 @@ def _get_input_file(self, given_path):
...
         :param dict config_entry: The scheme file entry
-        :param PeerContainer peer_container: The network topology of the cluster
+        :param PeerContainer peer_container_global: The global network topology
         :return: A matching NetworkConfig object
         :rtype: NetworkConfig
         """
-        self.check_keys_are_legal(config_entry, 'networkPolicyList', {'name': 1, 'networkPolicyList': 1,
-                                                                      'expected_warnings': 0})
+        self.check_keys_are_legal(config_entry, 'networkPolicyList', {'name': 1, 'namespaceList': 0, 'podList': 0,
+                                                                      'networkPolicyList': 1, 'expected_warnings': 0})
```
Comment (zivnevo) [inline network-config-analyzer/SchemeRunner.py:60]:
```suggestion
        self.check_keys_are_legal(config_entry, 'networkConfig', {'name': 1, 'namespaceList': 0, 'podList': 0,
                                                                      'networkPolicyList': 1, 'expected_warnings': 0})
```

File: docs/SchemeFileFormat.md [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 A scheme file is a yaml file, specifying what should be checked.
-It should contain at least the `networkPolicyLists` and the `queries` fields.
+It should contain at least the `networkConfigList` and the `queries` fields.
 
 | Field | Description | Value | Default |
 |-------|-------------|-------|---------|
 |namespaceList|A yaml/json file containing a list of namespaces in the cluster|file|Cluster namespaces|
 |podList|A json yaml/file containing a list of pods in the cluster|file|Cluster pods|
-|networkPolicyLists|Sets of NetworkPolicy to reason about|list of [NetworkPolicyList](#NetworkPolicyListobject) objects|
+|networkConfigList|A set of network configurations and policies to reason about|configuration files and list of [NetworkPolicyList](#NetworkPolicyListobject) objects|
```
Comment (zivnevo) [inline docs/SchemeFileFormat.md:9]: The Value of `NetworkConfigList` should be "list of NetworkConfig objects".
Then, rename `NetworkPolicyList` to `NetworkConfig` and update its possible fields.

File: network-config-analyzer/SchemeRunner.py (hunk 4 above)
Comment (zivnevo) [inline network-config-analyzer/SchemeRunner.py:60]:
```suggestion
        self.check_keys_are_legal(config_entry, 'networkConfig', {'name': 1, 'namespaceList': 0, 'podList': 0,
                                                                      'networkPolicyList': 1, 'expected_warnings': 0})
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: docs/SchemeFileFormat.md (hunk 5 above)
Comment (shai-doron) [reply docs/SchemeFileFormat.md]: done

File: network-config-analyzer/SchemeRunner.py (hunk 4 above)
Comment (shai-doron) [reply network-config-analyzer/SchemeRunner.py]: done

File: docs/SchemeFileFormat.md [hunk 6]
```diff
@@ -1,26 +1,30 @@
...
-It should contain at least the `networkPolicyLists` and the `queries` fields.
+It should contain at least the `networkConfigList` and the `queries` fields.
 
 | Field | Description | Value | Default |
 |-------|-------------|-------|---------|
-|namespaceList|A yaml/json file containing a list of namespaces in the cluster|file|Cluster namespaces|
-|podList|A json yaml/file containing a list of pods in the cluster|file|Cluster pods|
-|networkPolicyLists|Sets of NetworkPolicy to reason about|list of [NetworkPolicyList](#NetworkPolicyListobject) objects|
+|namespaceList|A global list of namespaces in the cluster|directory, git-repo or yaml/json file|Cluster namespaces|
+|podList|A global list of pods in the cluster|directory, git-repo or yaml/json file|Cluster pods|
```
Comment (zivnevo) [inline docs/SchemeFileFormat.md:8]:
```suggestion
|namespaceList|A globally-scoped list of namespaces in the cluster|directory, git-repo or yaml/json file|Cluster namespaces|
|podList|A globally-scoped list of pods in the cluster|directory, git-repo or yaml/json file|Cluster pods|
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -13,30 +13,30 @@ See an example scheme file [here](tests/example_policies/testcase1/testcase1-sch
...
+Checking whether the base set of NetworkConfigs permits the traffic explicitly specified in the given set of NetworkConfigs
+- `--forbids <NetworkConfig set> [--base_np_list <NetworkConfig set>]`\
+Checking whether the base set of NetworkConfigs forbids the traffic explicitly specified in the given set of NetworkConfigs
+
+`<NetworkConfig set>` should be one of:
+- a path to a yaml/json file defining NetworkConfigs
+- a path to a directory with files containing NetworkConfigs
+- a url of a GHE repository/dir/file with NetworkConfigs
+- The string `k8s`, instructing the tool to take all NetworkConfigs from a Kubernetes cluster (using `kubectl`)
+- The string `calico`, instructing the tool to take all NetworkConfigs from a Calico cluster (using `calicoctl`)
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply README.md:33]: ok, reverted

File: docs/SchemeFileFormat.md [hunk 8]
```diff
@@ -1,26 +1,30 @@
...
+|networkConfigList|A list of network configurations and policies to reason about|list of [NetworkConfig](#NetworkConfigobject) objects|
 |queries|Queries for the tool to run|list of [Query](#queryobject) objects|
 
-### <a name="NetworkPolicyListobject"></a>NetworkPolicyList object
-Each NetworkPolicyList object represents a set of NetworkPolicy objects (and possibly Calico's GlobalNetworkPolicy and Profile objects). It must contain the following two fields.
+### <a name="NetworkConfigobject"></a>NetworkConfig object
+Each NetworkConfig object represents a specific network configuration (and possibly Calico's GlobalNetworkPolicy and Profile objects).
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply docs/SchemeFileFormat.md:13]: done

File: docs/SchemeFileFormat.md (hunk 6 above)
Comment (shai-doron) [reply docs/SchemeFileFormat.md]: done

**Review by zivnevo — APPROVED** [review]
Great work

**Pull request merged**